File: work/crud-dotnet/Controllers/ItemsController.cs
using CrudDotNet.Models;
using CrudDotNet.Services;
using Microsoft.AspNetCore.Mvc;

namespace CrudDotNet.Controllers;

[Route("items")]
public class ItemsController : ControllerBase
{
    private readonly IItemRepository _items;

    public ItemsController(IItemRepository items)
    {
        _items = items;
    }

    [HttpGet]
    public IActionResult GetAll() => Ok(_items.GetAll());

    [HttpGet("{id:int}")]
    public IActionResult GetById(int id)
    {
        var item = _items.GetById(id);
        return item is null ? NotFound(new { error = "Item not found" }) : Ok(item);
    }

    [HttpPost]
    public IActionResult Create([FromBody] ItemInput? input)
    {
        if (input is null || string.IsNullOrWhiteSpace(input.Name))
            return BadRequest(new { error = "Field 'name' is required (string)" });

        var created = _items.Create(input);
        return StatusCode(StatusCodes.Status201Created, created);
    }

    [HttpPut("{id:int}")]
    public IActionResult Update(int id, [FromBody] ItemInput? input)
    {
        if (input?.Name is not null && string.IsNullOrWhiteSpace(input.Name))
            return BadRequest(new { error = "Field 'name' must be a non-empty string" });

        var updated = _items.Update(id, input ?? new ItemInput());
        return updated is null ? NotFound(new { error = "Item not found" }) : Ok(updated);
    }

    [HttpDelete("{id:int}")]
    public IActionResult Delete(int id)
    {
        var removed = _items.Delete(id);
        return removed is null ? NotFound(new { error = "Item not found" }) : Ok(removed);
    }
}

File: work/crud-nodejs/controllers/itemsController.js
const itemModel = require("../models/itemModel");
const { sendJson, parseBody } = require("../utils/httpHelpers");

async function getAll(req, res) {
  sendJson(res, 200, await itemModel.findAll());
}

async function getById(req, res, id) {
  const item = await itemModel.findById(id);
  if (!item) {
    sendJson(res, 404, { error: "Item not found" });
    return;
  }
  sendJson(res, 200, item);
}

async function create(req, res) {
  const body = await parseBody(req);
  if (!body.name || typeof body.name !== "string") {
    sendJson(res, 400, { error: "Field 'name' is required (string)" });
    return;
  }
  const item = await itemModel.create(body);
  sendJson(res, 201, item);
}

async function update(req, res, id) {
  const body = await parseBody(req);
  const existing = await itemModel.findById(id);
  if (!existing) {
    sendJson(res, 404, { error: "Item not found" });
    return;
  }
  if (body.name !== undefined) {
    if (typeof body.name !== "string" || !body.name.trim()) {
      sendJson(res, 400, { error: "Field 'name' must be a non-empty string" });
      return;
    }
  }
  const item = await itemModel.update(id, body);
  sendJson(res, 200, item);
}

async function remove(req, res, id) {
  const item = await itemModel.remove(id);
  if (!item) {
    sendJson(res, 404, { error: "Item not found" });
    return;
  }
  sendJson(res, 200, item);
}

module.exports = {
  getAll,
  getById,
  create,
  update,
  remove,
};
